Completely Rebuild our CI System

Myoldmopar opened the pull request:
Pull request overview
---------------------

## Original Comment 
Mac CI acted _crazy_ the last two days, and I can't figure out why.  I'm considering just moving it to GHA.  As a quick first step, I want to add basic build/test.  If that goes in smoothly, I may spend some time getting regressions working on GHA as well, using caching to store `develop` results for efficiency.  If this goes smoothly, this would **_rapidly_** eliminate Decent CI...

For now, just a test.

## Updated
Alright, so I completely rebuilt our CI system around GitHub Actions, and am going to throw Decent CI aside.  I'll give a full walkthrough of my changes here, but here are some highlights:
 - Regressions now show back up on the pull request page, not hidden on the commit
 - No more waiting on me to turn the CI machine back on if it goes down for updates
 - No more having to build and deploy custom machines at NREL
 - Super transparent build/test processes documented directly in the action workflow files
 - Running tests in a fully isolated environment
 - No more waiting on our CI machine to run one build at a time
 - No more Ruby code in my life, and no more managing our fork of Decent CI
 - I added regression testing across all three platforms again
 - No more managing the build results repository and page build issues
 - Nicely handle the standardized `[ ci skip ]` types of commit message [hints](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/skipping-workflow-runs).
 - Our testing focuses on open PRs, not just stray branches.
 - You can focus on the "Checks" tab at the top of the PR for full info about allll of our builds and tests
 - Probably lots more.

Anyway, this is very close to being ready.  I just pushed a significant change that may need a tweak or two, but still it's close.  For what it's worth, this doesn't touch any of our build processes for release packages, just testing.

The one piece I think is missing is the performance metrics.  I can add another workflow for that, but I'd love to discuss the best path forward to make it really useful.

Myoldmopar committed:
Add new Mac build/test GHA workflow

Myoldmopar committed:
Temporarily rename DCI file while playing with Mac workflow file

Myoldmopar committed:
Fixup test name and add build_testing

Myoldmopar committed:
Add upterm to debug failing tests

Myoldmopar committed:
Hacky regression stuff, lets see how close we are

Myoldmopar committed:
Try removing braces

Myoldmopar committed:
Use correct arg for runtests

Myoldmopar committed:
Add a little debugging output

Myoldmopar committed:
Rework mac build file, trying on all platforms just to see what happens

Myoldmopar committed:
Try again

Myoldmopar committed:
Cleanup on aisle build_and_test

Myoldmopar committed:
Be a little nicer with regression output

Myoldmopar committed:
Use ReadVars project name on windows

Myoldmopar committed:
Use | multiline yaml

Myoldmopar committed:
Use proper conditional for ubuntu FC assignment

Myoldmopar committed:
Add -C release for ctest

Myoldmopar committed:
Bump python version for 24.04 support

Myoldmopar committed:
Minor clean in new yaml, delete unused yamls

Myoldmopar committed:
Test only for now, lots of changes

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into MacBuildsOnGHA

Myoldmopar committed:
Nicer messaging, add a diff!

Myoldmopar committed:
Fix path issue, just trying on Mac14/U24

Myoldmopar committed:
Allow PR permission

Myoldmopar committed:
Re-enable all platforms

Myoldmopar committed:
Much improved regression page

Myoldmopar committed:
Clang format

Myoldmopar committed:
Add markdown summary to workflow page

Myoldmopar committed:
Just macos14 for now

Myoldmopar committed:
Add debug coverage first attempt

Myoldmopar committed:
Add coverage job summary, only run regressions on one per platform

Myoldmopar committed:
Remove Decent CI stuff, add develop build, cleanup alternate build

Myoldmopar commented on .decent_ci-Linux.yaml:
Deleted all four Decent CI files, gone.

Myoldmopar commented on .github/disabled-workflows/mac_test.yml:
Clean out a bunch of unused/unfinished workflows.

Myoldmopar commented on .github/workflows/alternate_linux_build.yml:
This is actually just a rename of the linux_build.yml to clarify the intention and an aggressive cleanup.  It's possible I need to tweak something, we'll see how it comes back in the next result.  But basically, this workflow's job is simple: configure and make a variant version of EnergyPlus using all the non-standard preprocessor DEFs.

Myoldmopar commented on .github/workflows/build_and_test_develop.yml:
This workflow is set up to just run on develop alone.  It builds on the full array of platforms, and uses the preprocessor DEF variables that we use in our release packages.  It then runs the test suite and leaves.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
This is where we are actually testing modifications.  For now I am running this for PRs only.  Devs are welcome to keep code in branches when it's not ready for any testing.  Once they want testing done, open a PR and mark it draft.  Once done mark it ready for review.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
```diff
@@ -0,0 +1,195 @@
+          echo "FC=gfortran-13" >> $GITHUB_ENV
+        fi
+
+    # BUILD AND TEST INTEGRATION FILES ON THE BASELINE BRANCH
```
For build types that we do regressions, this section checks out develop, builds it, and runs integrations.  It is possible we try to cache the results of this, I'm open to suggestion.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
```diff
@@ -0,0 +1,195 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
```
Full matrix of runs, note that one of each platform type has `run_regressions: true`.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
```diff
@@ -0,0 +1,195 @@
+
+    # BUILD AND TEST EVERYTHING ON THE CURRENT BRANCH
+
+    - name: Branch Checkout
```
Regardless of whether we are doing regressions, we will always build the branch and run the full test suite on it.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
```diff
@@ -0,0 +1,195 @@
+    - name: Branch Test
+      working-directory: ./branch/build
+      run: ctest -C Release -E Basement -j 3
+
```
Next, if we are doing regressions, we will install the regression tool python package and run our new script to process results.

Myoldmopar commented on .github/workflows/build_and_test_pr.yml:
```diff
@@ -0,0 +1,195 @@
+      if: always() && matrix.run_regressions && steps.branch_build.outcome != 'failure'  # always run this step as long as we actually built
+      id: regressions
+      run: python ./branch/scripts/dev/gha_regressions.py ./baseline/build/testfiles ./branch/build/testfiles/ ${{ github.workspace }}/regressions
+
```
And finally we report back to the dev in 3 ways:
 - Generating an artifact summary on the workflow run page that quickly summarizes the regression outcome.  This may not be necessary.  
![image](https://github.com/user-attachments/assets/286fc658-7122-4c7f-814e-2fe04680dcd8)
 - Uploading the full regression bundle as an artifact.  This will come down as a set of web pages in a zip.  Just extract it and open index.html, and enjoy all the diff information.  It's possible we could upload more than this...perhaps a Jupyter Notebook prepopulated with data for analyzing the diffs?
 - Finally we will also report back to the PR page to alert the dev of regressions.  If we choose to run on more than just PRs, then this will need to be modified to comment properly.  If you look at my commits here, you will see I tried to comment on commits, but realized it's probably not that important.

jmarrec replied:
Where is the artifact summary now? I can't see it

jmarrec replied:
I see the Coverage one here: https://github.com/NREL/EnergyPlus/actions/runs/10623481655?pr=10683

![image](https://github.com/user-attachments/assets/ad94b425-7b74-4875-8039-143612776da4)

jmarrec replied:
Ok, I see that the gha_regression.py isn't called for the Debug Integration workflows, and that the test_pull_requests creates one only if something failed.

jmarrec replied:
yet this one failed and I don't see a GITHUB_STEP_SUMMARY here https://github.com/NREL/EnergyPlus/actions/runs/10620499802

Myoldmopar commented on .github/workflows/debug_coverage.yml:
This file is where we are running EnergyPlus in debug and capturing code coverage results.  For now I'm lumping all tests together, not splitting it into unit/integration results.  We could add that, but I'm not sure if anyone is concerned enough to do it.

Myoldmopar commented on .github/workflows/debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+
+on:
+  pull_request:
+    branches: [ develop ]  # run this on any PR pointing to develop
```
- [x] TODO: This should also run on develop commits

Myoldmopar commented on .github/workflows/debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+    - name: Run Tests
+      working-directory: ./build
+      run: ctest -E Basement -j 4
+
```
Once you've built with ENABLE_COVERAGE and running tests, we simply lcov it to gather results and genhtml it to make an HTML bundle.

Myoldmopar commented on .github/workflows/debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+
+    - name: Generate HTML Coverage Results
+      working-directory: ./build
+      run: genhtml ./lcov.output.filtered -o lcov-html --demangle-cpp --function-coverage | tee cover.txt
```
- ~~TODO: Add -rc high and low values for coverage warnings/errors~~

Myoldmopar replied:
Nevermind, the defaults are fine.

Myoldmopar commented on .github/workflows/debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+
+    - name: Run Tests
+      working-directory: ./build
+      run: ctest -E Basement -j 4
```
If we chose to split integration/debug, we would need to filter down the tests on this line, gather results, clean everything, and redo it separately.  We'd end up with two bundles from the same build, which is nice.

Myoldmopar commented on .github/workflows/debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+      run: echo "$(cat ${{ github.workspace }}/build/cover.md)" >> $GITHUB_STEP_SUMMARY
+
+    - uses: actions/upload-artifact@v4
+      id: upload_regressions
```
- [x] TODO: This isn't called upload_regressions.

Myoldmopar commented on .gitlab-ci.yml:
Based on the contents of this file, I don't think anyone has used this in a very long time.

Myoldmopar commented on CMakeLists.txt:
```diff
@@ -200,52 +200,12 @@ else()
 endif()
 
 if(BUILD_TESTING)
-  option(ENABLE_REGRESSION_TESTING "Enable Regression Tests" OFF)
```
Deleted tons of things related to regression testing!

Myoldmopar commented on requirements.txt:
```diff
@@ -1,2 +1,6 @@
+
+# requirements for the CI regression testing scripts
+energyplus-regressions
+boto3
```
- [ ] TODO: We don't need boto3!

Myoldmopar commented on scripts/dev/gha_coverage_summary.py:
This file's job is to simply take the coverage command output and process it into a nice little summary.

Myoldmopar commented on scripts/dev/gha_regressions.py:
This file's job is to take baseline and mod folders, and call the regression tool on each test case found.  Then gather those results and present them in a really nice HTML package.

Myoldmopar commented on src/EnergyPlus/ConvectionCoefficients.cc:
```diff
@@ -6298,8 +6298,8 @@ Real64 CalcClearRoof(EnergyPlusData &state,
-    if (Rex > 0.1) { // avoid zero and crazy small denominators
-        Real64 tmp = std::log(1.0 + GrLn / pow_2(Rex));
+    if (Rex > 0.1) {                                // avoid zero and crazy small denominators
+        Real64 tmp = std::log1p(GrLn / pow_2(Rex)); // std::log(1.0 + GrLn / pow_2(Rex));
```
This change is _purely_ to generate controlled diffs.  It will be reverted before this merges.

Myoldmopar commented on tst/EnergyPlus/unit/CommandLineInterface.unit.cc:
```diff
@@ -385,7 +385,7 @@ TEST_F(CommandLineInterfaceFixture, runReadVars)
 }
 
-TEST_F(CommandLineInterfaceFixture, numThread)
+TEST_F(CommandLineInterfaceFixture, DISABLED_numThread)
```
@jmarrec this test failed on me for mac 14 on GHA IIRC.  Do you know why this would fail?  I can provide more context later, but if this work is otherwise ready to go, I'm likely to not investigate the problem, and instead disable these temporarily.

Myoldmopar commented on tst/EnergyPlus/unit/Timer.unit.cc:
```diff
@@ -60,7 +60,7 @@
 using namespace EnergyPlus;
 
-TEST_F(EnergyPlusFixture, Timer_ticktock)
+TEST_F(EnergyPlusFixture, DISABLED_Timer_ticktock)
```
@jmarrec this test failed on me for mac 14 on GHA IIRC.  Do you know why this would fail?  I can provide more context later, but if this work is otherwise ready to go, I'm likely to not investigate the problem, and instead disable these temporarily.

Myoldmopar reviewed (commented):
OK, full walkthrough done, with a few TODO's that I'll tackle now.  I'm letting it run one more time to get full results and then I'll show it off on our next call.

Myoldmopar committed:
Add license to new py file

Myoldmopar committed:
Run debug coverage on develop

Myoldmopar committed:
Clean typo name and remove boto3 from reqs

Myoldmopar commented:
TODO's all addressed, just staged locally while I wait on updated results.

Myoldmopar committed:
Fix Git issue for missing submodule

Myoldmopar committed:
Try using env for summary

Myoldmopar committed:
Clean up regression output

Myoldmopar committed:
Only run coverage on unit tests for now

Myoldmopar committed:
Try fixing windows; add integration coverage back

Myoldmopar committed:
Split unit test coverage and integration debug builds

Myoldmopar committed:
Pull in JS attempt here

Myoldmopar committed:
Fixup naming of testing for debug runs

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into MacBuildsOnGHA

Myoldmopar committed:
Add license to python

Myoldmopar committed:
Proper matrix variable lookup, smh, try adding a warning annotation to diffs

Myoldmopar committed:
Add threading back in to debug builds

Myoldmopar committed:
Move targets out of matrix, clean up if conditions

Myoldmopar committed:
Ignore build dir assets when generating coverage results

Myoldmopar committed:
Set default shell to bash

Myoldmopar committed:
Remove debug message

Myoldmopar committed:
Remove windows-2019, move alternate build to just develop commits

Myoldmopar committed:
Workflow minor cleanups

Myoldmopar committed:
Try to fix windows path

Myoldmopar committed:
Just turn off regressions on Windows

Myoldmopar committed:
Disable regressions on u24, solely mac now, and skip the longest running debug integration tests

Myoldmopar committed:
Actually exclude files properly

Myoldmopar committed:
Move cppcheck, custom_check, and clang_format into a single job for CIs sake

Myoldmopar committed:
Comment on PR workflow

Myoldmopar committed:
Fix cppcheck alignment

Myoldmopar committed:
Fix workflow triggers

Myoldmopar committed:
Remove doc build windows workflow, pull ops into core documentation workflow

Myoldmopar committed:
Rename workflows

Myoldmopar committed:
Try removing the auto-add PR workflow, it seems to be unnecessary now

Myoldmopar committed:
Purely renaming workflow jobs so they make sense on the GitHub UI

Myoldmopar committed:
Fix documentation artifact uploads

Myoldmopar committed:
Remove line causing diffs

Myoldmopar commented:
OK, I took out the line causing diffs, trimmed back the number of runners and number of jobs as much as I feel comfortable for now, and cleaned up naming/etc.  Assuming this is clean, I am done here.  If this needs to sit for a spell, that's OK I guess, but we could also just revert the PR if we find an issue.  I won't like...destroy the Decent CI machines _right away_ when this merges.

Myoldmopar committed:
Only upload doc PDFs on Linux

Myoldmopar committed:
Minor renaming

Myoldmopar commented:
FWIW, when a commit is made to a PR now, 13 runners start up:
 - 3 EpJSON dependency checks -- @jmarrec could we pull this back to 1 Linux?
 - 2 Documentation: 1 Unix and 1 Windows because they are vastly different experiences
 - 1 Code Integrity runner to run cppcheck, clang format, and custom check
 - 1 PR label verification
 - 4 Unit/Regression Tests
   - Mac x64
   - Mac arm
   - Ubuntu 24.04
   - Windows
 - 2 Debug/Coverage runs

Of these 13, the 3 EpJSON and 1 Label verification are done in like 1 minute.  The documentation builds take less than 10 minutes.   Code integrity check takes about 25.  That leaves the 6 test builds, and I've sped them up heavily so that 5 of them are done in about an hour or less.  The integration debug build is still the slowest and is over 2 hours, but all the rest are done way before that.  

I'm pretty happy with this, but I could imagine one change for times sake.  If we want to add an extra runner, I could split up the debug integration runs into two separate jobs, with one test running the first 400, and the second running whatever is left.  That would get the entire CI suite done in probably less than 90 minutes.  I don't think it's worth it, but I'm open to your thoughts.

amirroth commented:
Most of the regressions are run with `Timestep, 6`. Is that resolution really necessary for all of them? For any of them?

Myoldmopar commented:
That is a fantastic idea!  It would be super easy to make it TimeStep 2 during all the debug builds at least.  I think I'll investigate....

Myoldmopar committed:
Add env var CI_FORCE_TIME_STEP=Y to force 30 minute time steps

mjwitte commented:
> That is a fantastic idea! It would be super easy to make it TimeStep 2 during all the debug builds at least. I think I'll investigate....

If you stick to debug builds, then I won't object. TimeStep = 6 *is* the default. That's a reason to keep it for regressions, maybe not a great reason, but it's a reason.

Myoldmopar committed:
Add new env var to the GitHub debug integration workflow

Myoldmopar commented:
```
$ ctest -R HospitalBaselineReheat
1/1 Test #386: integration.HospitalBaselineReheatReportEMS ...   Passed    8.47 sec

$ export CI_FORCE_TIME_STEP=Y

$ ctest -R HospitalBaselineReheat
1/1 Test #386: integration.HospitalBaselineReheatReportEMS ...   Passed    4.26 sec
```

Yep, definitely doing this for debug builds...  thanks @amirroth !!


And while I was commenting I went ahead and did it.  I'm thinking CI will be clean and we'll be getting super efficient!!

Myoldmopar commented:
OK, that's chugging along, we'll see how it goes with that super quick hack.

I'd love to hear more thoughts on speeding up our testing.  On Ubuntu, Windows, and Mac x64, it's a surprisingly long 35 minute build time, followed by a quick-ish 10 minute run of the test suite.  Mac arm64 is brutally fast relative to that.  

I am also open to being a little more aggressive on caching, but I'm only doing regressions on Mac arm64, and the entire baseline build/test is only 11 minutes.  I guess if we just named the baseline build directory like baseline-sha123abc, then it would be easy enough to cache it and see if it's already done.

Myoldmopar commented:
Quick check on the debug integration CI build.  Before:

```
 62/787 Test  #21: integration.VSHeatPumpWaterToAirWithRHControl .....   Passed  119.82 sec
```

With the timestep suggestion from @amirroth:

```
 61/787 Test  #36: integration.1ZoneUncontrolledFourAlgorithms .......   Passed   33.35 sec
```

Yeeeeee!!

amirroth commented:
Are you not multi-threading the builds? My 12 year old laptop builds EnergyPlus faster than that.

Myoldmopar commented:
Oh yes, they are threaded.  But the CI instances only have 3 or 4 logical cores depending on platform.

amirroth commented:
I just need to remove more code from EnergyPlus. That may be the only way.

Myoldmopar committed:
Fix failing integration test on debug

amirroth commented:
> I just need to remove more code from EnergyPlus. That may be the only way.

20% is a conservative estimate. 40% is possible.

jmarrec commented on .github/workflows/test_debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+      run: >
+        cmake 
+        -G "Unix Makefiles"
+        -DCMAKE_BUILD_TYPE:STRING=Debug
```
To speed up this workflow, you should try doing `-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -DFORCE_DEBUG_ARITHM_GCC_OR_CLANG:BOOL=ON`.

You should still catch NaNs and co, and have coverage working

Myoldmopar replied:
I was worried that this would mute NaNs, but if that's not a problem, then I'll just do that right now.  That will be a fantastic speed up.

jmarrec commented on .github/workflows/test_code_integrity.yml:
```diff
@@ -0,0 +1,104 @@
+        uses: jidicula/clang-format-action@v4.13.0
+        if: always()
+        with:
+          clang-format-version: '10'
```
Side note but it'd be time to live in the present. LLVM is at version 18 right now

Myoldmopar replied:
I can do this, but I will wait until a different PR.  Devs may have 10 specifically set up to auto format locally on save, and we should coordinate a moment to change it.  In case it makes a bunch of tweaks to the code, we may want to properly approach that and add it to the Git skip commit list.

jmarrec commented on .github/workflows/test_linux_alternate.yml:
```diff
@@ -0,0 +1,40 @@
+
+      - name: Configure CMake
+        working-directory: ./build/
+        run: >
```
Weird that you chose to use `>` and not `|` like the rest

Myoldmopar replied:
Not weird, superior.

jmarrec commented on .github/workflows/test_debug_integration.yml:
```diff
@@ -0,0 +1,83 @@
+      run: >
+        cmake 
+        -G "Unix Makefiles"
+        -DCMAKE_BUILD_TYPE:STRING=Debug
```
Same point RelWithDeb and FORCE_DEBUG_ARITHM_GCC_OR_CLANG ?

jmarrec commented on .github/workflows/test_develop_commits.yml:
```diff
@@ -0,0 +1,108 @@
+          arch: x86_64
+          python-arch: x64
+          generator: "Unix Makefiles"
+          nproc: 3
```
A bit annoying to harcode that, you will definitely end up outdated and not realize it for a long time (I had this happen already on OpenStudio).

brew coreutils to get nproc, or use `echo "N=$(nproc 2>/dev/null || sysctl -n hw.logicalcpu)" >> $GITHUB_ENV` ?

Myoldmopar replied:
I think if you look back at the commit log here, you'll see I tried to do that dynamically but ran into issues.  After all the time I've spent in here the last week I could probably re-enable that.

jmarrec commented on .github/workflows/test_develop_commits.yml:
```diff
@@ -0,0 +1,108 @@
+        cmake 
+        -G "${{ matrix.generator }}"
+        -DCMAKE_BUILD_TYPE:STRING=Release
+        -DCMAKE_OSX_DEPLOYMENT_TARGET:STRING=${{ matrix.macos_dev_target }}
```
(Surprised this works on non macOS)

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+    - name: Baseline Configure CMake
+      if: matrix.run_regressions
+      working-directory: ./baseline/build
+      run: >
+        cmake 
```
Again the `>` and there are a still of trailing whitespaces in that file, including on the "cmake " line.

Myoldmopar replied:
I will remove trailing white space, but I'm leaving > because it's so superior.

jmarrec commented on .github/workflows/test_develop_commits.yml:
```diff
@@ -0,0 +1,108 @@
+      working-directory: ./build
+      run: cmake --build . -j ${{ matrix.nproc }} --config Release
+
+    - name: Test
+      working-directory: ./build
+      run: ctest -C Release -E Basement -j ${{ matrix.nproc }}
```
Ok, so on push to develop, you run the tests, including the integration tests.

What you could do here is to cache the results.

> [**Cache scopes**](https://github.com/actions/cache?tab=readme-ov-file#cache-scopes)

> The cache is scoped to the key, [version](https://github.com/actions/cache?tab=readme-ov-file#cache-version), and branch. The default branch cache is available to other branches.

So you could cache with a key of `regression_results-${{ matrix.os }}-develop`

Myoldmopar replied:
Yep, that's what I was thinking it would be like.  I'll give it a try.

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+    - name: Baseline Test
+      if: matrix.run_regressions
+      working-directory: ./baseline/build
+      run: ctest -C Release -E Basement -R integration -j 3  # TODO: Speed up basement so we don't have to skip it.
```
That way, in the PR workflow, you do not have to checkout and build develop. You just access the latest develop results.

jmarrec commented on .github/workflows/test_debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+
+    - name: Build
+      working-directory: ./build
+      run: cmake --build . -j 4
```
General comment is that there is a LOT of repeated code in all of these workflows.

Using a composite action (just yaml) stored in-repo (in `.github/actions`) might be good to avoid changing 8 files at once. The cons is that it obfuscate a bit what's happening since you have to look at two files.

jmarrec replied:
Anywaaaaays, the integration testing one is doing:

`cmake --build . -j 4 --target energyplus ExpandObjects ReadVarsESO Slab Basement AppGPostProcess ParametricPreprocessor` and takes 14min instead of 32min. Can this be done here too?

Myoldmopar replied:
For this one we have to build `energyplus_tests`, but we probably don't need a bunch of the other things.  I'll update this with a specific list to trim down the build time.

jmarrec commented on .github/workflows/test_debug_integration.yml:
```diff
@@ -0,0 +1,83 @@
+        -DUSE_OpenMP:BOOL=OFF
+        ../
+
+    - name: Build
+      working-directory: ./build
+      run: cmake --build . -j 4 --target energyplus ExpandObjects ReadVarsESO Slab Basement AppGPostProcess ParametricPreprocessor
```
This takes "only" 14 min, so I don't recommend doing the following but I'm still going to mention it.

I have wondered if we instead couldn't merge the debug_coverage and debug_integration into the same workflow and do a build, then in parallel runners, run the coverage and integration.

```yml
jobs:
   build:
      steps: 
      - build
      - build TGZ package
      - upload-artifact 

  run_coverage:
      needs: build
      steps:
      - checkout
       - download-artifact
       - run coverage
  
   run_integration:
      steps:
       - checkout
       - download-artifact
       - run integration
 ```

Myoldmopar replied:
Definitely an option.  I'll hold off for right now.

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+      with:
+        path: branch
+
+    - name: Branch Create Build Directory
+      run: cmake -E make_directory ./branch/build/
+
+    - name: Branch Configure CMake
+      working-directory: ./branch/build
+      run: >
```
Setup CCache to speed up incremental?

Also, why is this taking 34 min? https://github.com/NREL/EnergyPlus/actions/runs/10623481639/job/29449990399?pr=10683

When the Debug build takes 32min: https://github.com/NREL/EnergyPlus/actions/runs/10623481655/job/29449989580?pr=10683

Myoldmopar replied:
I was hesitating on this because I didn't want to add more complication while figuring all this out.  But if using ccache is straightforward, then by all means we should enable it.  I'll look into it.

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+
+jobs:
+  build_and_test:
+    name: Unit/Regression Testing
```
I find this unreadable

![image](https://github.com/user-attachments/assets/df62cb2f-59dc-4d16-8725-e0cfb4ee6a69)


I suggest actually using a unique name here.

something like 

```suggestion
    name: Unit/Regression Testing ${{ matrix.pretty_name }}
```

Myoldmopar replied:
Makes great sense.  I hadn't thought of using a matrix variable in the name, good stuff.

jmarrec commented on .github/workflows/build_documentation.yml:
```diff
@@ -0,0 +1,226 @@
+
+jobs:
+  build:
+    name: Build and Upload PDF Documentation
```
```suggestion
    name: Documentation ${{ matrix.os }}
```

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+
+    - name: Generate Regression Summary GitHub Script
+      if: always() && matrix.run_regressions && steps.regressions.outcome == 'failure'
+      run: > 
+        python ./branch/scripts/dev/build_regression_summary.py 
+        ${{ github.workspace }}/regressions/summary.md
+        ${{ github.workspace }}/regressions/summary.js
+        ${{ matrix.os }}
+        ${{ github.sha }}
+        ${{ github.run_id }}
+        ${{ steps.upload_regressions.outputs.artifact-url }}
```
Oh wow, TIL that the `>` allows you to run a command without splittling long lines with ` \` I guess.

Myoldmopar replied:
So superior.

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+        ${{ github.run_id }}
+        ${{ steps.upload_regressions.outputs.artifact-url }}
+
+    - uses: actions/github-script@v7
+      if: always() && matrix.run_regressions && steps.regressions.outcome == 'failure'
+      with:
+        script: |
+          const script = require('${{ github.workspace }}/regressions/summary.js')
+          console.log(script({github, context}))
```
It'd be good to step the content of summary.md to the GITHUB_STEP_SUMMARY as well.

and if everything passes, maybe consider still writing a summary, such a number of tests run and 100% pass rate in green or something. Just nitpicking here

jmarrec commented on .github/workflows/test_pull_requests.yml:
```diff
@@ -0,0 +1,199 @@
+
+    - name: Branch Test
+      working-directory: ./branch/build
+      run: ctest -C Release -E Basement -j 3
```
Ideally we'd process the ctest JUnit results, and create **annotations** on the workflow for any test that failed.

Myoldmopar replied:
Yeah, it would be great if there was a script already available that could process the ctest results.  If there is, I'm open to bringing it in and using it.  But developing that would probably take a good several hours from an experienced dev.

jmarrec commented on .github/workflows/test_debug_coverage.yml:
```diff
@@ -0,0 +1,89 @@
+      working-directory: ./build
+      run: genhtml ./lcov.output.filtered -o lcov-html --demangle-cpp --function-coverage | tee cover.txt
+
+    - name: Process Unit Test Coverage Summary
+      working-directory: ./build
+      run: python ${{ github.workspace }}/scripts/dev/gha_coverage_summary.py
+
+    - name: Generate Unit Test Artifact Summary
+      run: echo "$(cat ${{ github.workspace }}/build/cover.md)" >> $GITHUB_STEP_SUMMARY
```
Nit: why use two steps? Just have the gha_coverage_summary.py write to GITHUB_STEP_SUMMARY directly.

You could do stuff like


```
print(content)
if 'GITHUB_ACTIONS' in os.environ:
    with open(os.environ('GITHUB_STEP_SUMMARY'), 'a') as f:
         f.write(content)
```

so one could run it locally and see it.

Similar to what I'm doing here: https://github.com/openstudiocoalition/OpenStudioApplication/blob/a78a4434a1cfe891f054ff048d4cd5ed4b1889c5/ci/parse_cmake_versions.py#L55-L74

Myoldmopar replied:
Yeah, the GITHUB_STEP_SUMMARY crap took way too many iterations to get it working well.  I'm not likely to pull on that thread for right now, but **totally** open to tweaking it later or accepting changes to it if they are offered.

Myoldmopar committed:
Take out macos-12 from each PR commit build, we will still build on develop and provide packages for mac x64

Myoldmopar committed:
Set build type to RelWithDebInfo and force debug arithmetic on debug builds

Myoldmopar committed:
Explicitly list targets for debug coverage and turn off fortran, remove trailing spaces in workflows

Myoldmopar committed:
Adding nicer names for the workflows and jobs, open to suggestion

Myoldmopar commented:
OK, I just pushed a handful of commits that further clean things up.  Key aspects are:
 - Trying RelWithDebInfo for both debug builds, which should drastically clean things up
 - Removing macos-12 from regular PR testing, we don't need to waste those cycles for every commit, and we are still building it for commits to develop and release packages
 - I'm trying to explicitly list the build targets for the debug coverage run, should also help save time
 - I tried to clean up the names of the workflows and jobs.  I think it looks a lot better in the list, but I'm open to further refinement!

Things I did not do yet:
 - Add any caching, whether using GHA caching, or ccache itself.  I just didn't want to introduce too many changes in one set of GHA runs
 - Dynamically set NPROC, just to again avoid breaking something random with all my other changes. 
 - Make any changes to the GITHUB_STEP_SUMMARY or ctest output.  Still open to these, whether in this PR or later.

Myoldmopar commented:
That naming is pretty nice looking!  I'm definitely open to further refinement, but this is great.  

![image](https://github.com/user-attachments/assets/ae3c64f8-9e1c-4831-8bd5-a938687c047f)

There is also the `workflow.run_name` attribute that we could add, but I'm not sure exactly where that is shown on the GitHub UI, so maybe it's not relevant to us.

Myoldmopar commented:
I think I actually _will_ bring the debug builds together in the same workflow file at least.  I'm not sure if I'll do the artifact upload/download just yet, because I think we may have each compilation split up sufficiently that they are pretty quick now.  But either way, putting them in the same workflow will make the UI look that much nicer.

Myoldmopar committed:
Bring linux alternate into the main develop commit check file

Myoldmopar committed:
Bring debug builds into the same workflow file

Myoldmopar committed:
Skip SSC file for NaN

Myoldmopar commented:
- Unit test coverage: 1 hour
- Integration debug: 38 minutes!  (1 test failure due to NaNs in SSC)
- 3 unit/regressions: all done in less than an hour

Others all done quickly.  This is looking super great.  I am now pushing one more time to skip that SSC file, and also regroup our related CI builds together.  I don't _think_ it should affect anything, but if so I'll fix it up with one more commit and I think we're safe to merge after that.  Follow-up work would include fixing up caching, NPROC, parsing the ctest results, and the GITHUB_STEP_SUMMARY stuff.  But this doesn't need to wait on those.

Myoldmopar commented:
So it seems the RelWithDebInfo cmake build type is actually muting debug assertions, because it is still definiing NDEBUG.  I think what I'd like to do is just modify our Cmake build commands so that it removes NDEBUG during RelWithDebInfo builds.  I think nearly no one ever builds with that type, so it won't affect anything.  And regardless it seems reasonable to include assertions if you need debug info...

Thoughts @jmarrec ?

amirroth commented:
I build with it fairly often (certain issues don't reproduce in debug builds), but I also muck around with the Makefiles locally so you can do whatever you need to for CI purposes.

jmarrec commented:
> So it seems the RelWithDebInfo cmake build type is actually muting debug assertions, because it is still definiing NDEBUG. I think what I'd like to do is just modify our Cmake build commands so that it removes NDEBUG during RelWithDebInfo builds. I think nearly no one ever builds with that type, so it won't affect anything. And regardless it seems reasonable to include assertions if you need debug info...
> 
> Thoughts @jmarrec ?

https://github.com/Kitware/CMake/blob/9673f61be1573b7e2b701da8a6ad0fb4c8cac4a1/Modules/Compiler/GNU.cmake#L61-L65

Indeed it does. Yeah we could tweak our cmake config to remove it, or I think we could just do `-UNDEBUG`.

```
target_compile_options(project_options INTERFACE $<$<CONFIG:RelWithDebInfo>:-UNDEBUG>)
```

Myoldmopar committed:
Add -UNDEBUG for RelWithDebInfo on GCC, could add to Clang/VS later

Myoldmopar commented:
> I think we could just do `-UNDEBUG`

Thanks @jmarrec, that did the trick.  I added it specifically into the GCC compiler flag block so that it wouldn't affect anything else for now.  I added an `assert(false)` into `SimulationManager` and ran it with Debug, Release, and RelWithDebInfo.  It passed on release, of course, but asserted on both of the others.  This is perfect.  OK, this was the last thing holding this up.  I think this can merge at will if it comes back clean this morning.

As an upcoming step, I will add a nice way to report some metrics from the performance testing, but honestly, I'd like to get input from the team on what we'd like to see from that stuff.  We can probably do lots better than what we are doing now, and I don't want to waste time simply bringing over existing stuff just to throw it away.

Myoldmopar commented:
Another interesting change here.  It looks like GitHub is automatically checking out the /merge/ ref, which brings develop into the branch. This eliminates the issue of being behind develop and having false diffs.  So that's great.  However.....it may just not run if there are conflicts.  So that's not great.  I can look into this later, but I don't think this should wait any longer.  It just completed the entire suite of tests in an hour and is all green.  Anyone have any final thoughts here?  Note that this PR does bring in a few changes to develop branch testing as well, so I'm totally open to the possibility that I will need to quickly fix something once it merges.  But maybe it will all be fine? :shrug:

rraustad commented:
This will disable the existing CI system. Might be good to have the Linux CI active for a period of time to compare results?

Myoldmopar commented:
I don't hate that idea @rraustad, although I was already comparing diffs on this branch by temporarily including your log1p change.  

Decent CI:

![image](https://github.com/user-attachments/assets/57a4cb95-f611-4781-9e9e-65de0fca55ee)
![image](https://github.com/user-attachments/assets/40a3ef0b-af92-4a40-a788-030c59dad4f0)

The new CI regression diffs:

![image](https://github.com/user-attachments/assets/865fa2f8-9983-46fd-b0b2-563be8a9ea89)

I think things are working really great.  If you feel strongly about it, I'm OK putting those Decent files back in, but if not, I'll lean away from it and see how things go.

Myoldmopar commented:
Oh there actually is a difference, but it's only because on the new CI I am skipping the long running Basement files, which includes the KivaBasement.  Now that the new CI stuff is working so much better, I could probably just add that back in, which will bring it into agreement with the existing Decent CI diffs.

Myoldmopar commented:
Tell you what.....  I'll do both for now.  I'm going to re-enable Basement tests and also add back in Decent CI's Linux and Windows stuff temporarily.

Myoldmopar committed:
Take out Basement file skipping

Myoldmopar commented:
Nope, scratch that as well.  I've taken out some CMake functionality that would need to go back in.  I am happy to help get that stuff added back into your fork when the time comes, but I'm just going to pull the plug on our Decent CI stuff for now.  Worst case for me is that this falls apart when I merge and I have to just revert the whole PR :)  That's all.  

I will take out the ctest argument that skips the Basement files so that our regressions will match exactly what Decent CI was producing.  And then I will merge this after lunch.

Myoldmopar commented:
OK, all super happy now.  Adding the basement files only added a couple minutes on it looks like.  I am going to merge this.  Let's see how it goes once in the develop branch, and with more branches running.

Myoldmopar commented:
WELLLLL after all that testing and no issues with queued up runners waiting, I merged it into develop, and immediately it's too much.  Geez.  OK, well, I'm reverting this to avoid breaking anything, and I'll look back into self-hosted runners.  Or whatever.
